File: src/SolitaireGame.js
/*
 * decaffeinate suggestions:
 * DS101: Remove unnecessary use of Array.from
 * DS102: Remove unnecessary code created because of implicit returns
 * DS205: Consider reworking code to avoid use of IIFEs
 * DS207: Consider shorter variations of null checks
 * Full docs: https://github.com/decaffeinate/decaffeinate/blob/main/docs/suggestions.md
 */
import * as cardutils from "./cardutils"

// -------------------------------------------------------------------------------------------------

class SolitaireGame {
    constructor() {
        this.saveTimeout = null
        this.state = null
        this.mode = "klondike"
        this.hard = false // this is the toggle for the *next* game. Check @state.hard to see if *this* game is hard
        this.undoStack = []
        this.tweens = []
        this.nextTweenID = 0

        this.modes = {}
        this.loadMode("baker")
        this.loadMode("eagle")
        this.loadMode("emperor")
        this.loadMode("freecell")
        this.loadMode("golf")
        this.loadMode("klondike")
        this.loadMode("scorpion")
        this.loadMode("spider")
        this.loadMode("spiderette")
        this.loadMode("yukon")
        this.loadMode("luckyseven")

        if (!this.load()) {
            this.newGame()
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Mode Manipulation

    loadMode(modeName) {
        const modeFuncs = {
            newGame: true,
            click: true,
            won: true,
            lost: true,
            phase: true
        }

        const modeProps = {
            name: true,
            help: true
        }

        const m = require(`./modes/${modeName}`).default
        this.modes[modeName] = {}

        return (() => {
            const result = []
            for (var key in m) {
                var val = m[key]
                if (modeFuncs[key]) {
                    result.push((this.modes[modeName][key] = val.bind(this)))
                } else if (modeProps[key]) {
                    result.push((this.modes[modeName][key] = val))
                } else {
                    result.push((this[key] = val.bind(this)))
                }
            }
            return result
        })()
    }

    // -----------------------------------------------------------------------------------------------
    // Generic input handlers

    load() {
        // return false

        let payload
        const rawPayload = localStorage.getItem("save")
        if (rawPayload == null) {
            return false
        }

        try {
            payload = JSON.parse(rawPayload)
        } catch (error) {
            return false
        }

        this.mode = payload.mode
        this.hard = payload.hard === true
        this.state = payload.state
        this.undoStack = []
        console.log("Loaded.")
        return true
    }

    save() {
        // return

        const payload = {
            mode: this.mode,
            hard: this.hard,
            state: this.state
        }
        return localStorage.setItem("save", JSON.stringify(payload))
    }
    // console.log "Saved."

    queueSave() {
        // Don't bother queueing, just save immediately
        return this.save()
    }

    // if @saveTimeout?
    //   clearTimeout(@saveTimeout)
    // @saveTimeout = setTimeout =>
    //   @save()
    //   @saveTimeout = null
    // , 3000

    // -----------------------------------------------------------------------------------------------
    // Undo

    canUndo() {
        return this.undoStack.length > 0
    }

    pushUndo() {
        return this.undoStack.push(JSON.stringify(this.state))
    }

    undo() {
        if (this.undoStack.length > 0) {
            this.state = JSON.parse(this.undoStack.pop())
            return this.queueSave()
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Generic input handlers

    newGame(newMode = null) {
        if (newMode != null && this.modes[newMode] != null) {
            this.mode = newMode
        }
        if (this.modes[this.mode] != null) {
            this.modes[this.mode].newGame()
            this.undoStack = []
            return this.save()
        }
    }

    click(type, outerIndex, innerIndex, isRightClick, isMouseUp) {
        if (outerIndex == null) {
            outerIndex = 0
        }
        if (innerIndex == null) {
            innerIndex = 0
        }
        if (isRightClick == null) {
            isRightClick = false
        }
        if (isMouseUp == null) {
            isMouseUp = false
        }
        console.log(`game.click(${type}, ${outerIndex}, ${innerIndex}, ${isRightClick}, ${isMouseUp})`)
        if (this.modes[this.mode] != null) {
            if (!isMouseUp) {
                this.pushUndo()
            }
            this.modes[this.mode].click(type, outerIndex, innerIndex, isRightClick, isMouseUp)
            return this.queueSave()
        }
    }

    won() {
        if (this.modes[this.mode] != null) {
            return this.modes[this.mode].won()
        }
        return false
    }

    lost() {
        if (this.modes[this.mode] != null && this.modes[this.mode].lost != null) {
            return this.modes[this.mode].lost()
        }
        return false
    }

    phase() {
        if (this.modes[this.mode] != null && this.modes[this.mode].phase != null) {
            this.modes[this.mode].phase()
            return this.queueSave()
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Generic helpers

    findFoundationSuitIndex(raw) {
        let f, fIndex
        if (this.state.foundations.length === 0) {
            return -1
        }

        // find a matching suit
        const srcInfo = cardutils.info(raw)
        for (fIndex = 0; fIndex < this.state.foundations.length; fIndex++) {
            f = this.state.foundations[fIndex]
            if (f >= 0) {
                var dstInfo = cardutils.info(f)
                if (
                    srcInfo.suit === dstInfo.suit &&
                    (srcInfo.value === dstInfo.value + 1 || srcInfo.value === dstInfo.value - 12)
                ) {
                    return fIndex
                }
            }
        }

        // find a free slot
        for (fIndex = 0; fIndex < this.state.foundations.length; fIndex++) {
            f = this.state.foundations[fIndex]
            if (f < 0) {
                return fIndex
            }
        }

        return -1
    }

    dumpCards(prefix, cards) {
        const infos = []
        for (var card of Array.from(cards)) {
            infos.push(cardutils.info(card))
        }
        return console.log(prefix, infos)
    }

    workPileEmpty() {
        for (var work of Array.from(this.state.work)) {
            if (work.length > 0) {
                return false
            }
        }
        return true
    }

    reserveEmpty() {
        if (this.state.reserve == null) {
            return true
        }
        for (var cols of Array.from(this.state.reserve.cols)) {
            if (cols.length > 0) {
                return false
            }
        }
        return true
    }

    canAutoWin() {
        let col
        if (this.won()) {
            // console.log "canAutoWin: false (won)"
            return false
        }
        if (this.state.foundations.length === 0) {
            // console.log "canAutoWin: false (no foundations)"
            return false
        }
        if (this.state.draw.cards.length > 0) {
            // console.log "canAutoWin: false (has draw)"
            return false
        }
        if (this.state.pile.cards.length > 0) {
            // console.log "canAutoWin: false (has pile)"
            return false
        }
        if (this.state.reserve != null) {
            for (col of Array.from(this.state.reserve.cols)) {
                if (col.length > 0) {
                    // console.log "canAutoWin: false (has reserves)"
                    return false
                }
            }
        }
        for (let colIndex = 0; colIndex < this.state.work.length; colIndex++) {
            col = this.state.work[colIndex]
            var last = 100
            for (var raw of Array.from(col)) {
                if (raw & cardutils.FLIP_FLAG) {
                    // console.log "canAutoWin: false (has flip)"
                    return false
                }
                var info = cardutils.info(raw)
                if (info.value > last) {
                    // console.log "canAutoWin: false (not descending #{colIndex})"
                    return false
                }
                last = info.value
            }
        }
        // console.log "canAutoWin: true"
        return true
    }

    sendHome(type, outerIndex, innerIndex) {
        let srcCol
        let src = null
        switch (type) {
            case "pile":
                if (this.state.pile.cards.length > 0) {
                    src = this.state.pile.cards[this.state.pile.cards.length - 1]
                }
                break
            case "reserve":
                if (
                    this.state.reserve != null &&
                    this.state.reserve.cols.length > outerIndex &&
                    this.state.reserve.cols[outerIndex].length > 0
                ) {
                    src = this.state.reserve.cols[outerIndex][this.state.reserve.cols[outerIndex].length - 1]
                }
                break
            case "work":
                srcCol = this.state.work[outerIndex]
                if (innerIndex !== srcCol.length - 1) {
                    return false
                }
                src = srcCol[innerIndex]
                break
        }

        if (src == null) {
            return null
        }

        const srcInfo = cardutils.info(src)
        const dstIndex = this.findFoundationSuitIndex(src)
        if (dstIndex >= 0) {
            let sendHome = false
            if (this.state.foundations[dstIndex] >= 0) {
                const dstInfo = cardutils.info(this.state.foundations[dstIndex])
                if (srcInfo.value === dstInfo.value + 1 || srcInfo.value === dstInfo.value - 12) {
                    sendHome = true
                }
            } else {
                let foundationBase = 0 // Ace
                if (this.state.foundationBase != null) {
                    ;({ foundationBase } = this.state)
                }
                if (srcInfo.value === foundationBase) {
                    sendHome = true
                }
            }

            if (sendHome) {
                const prevFoundationRaw = this.state.foundations[dstIndex]
                this.state.foundations[dstIndex] = src
                switch (type) {
                    case "pile":
                        this.state.pile.cards.pop()
                        break
                    case "reserve":
                        this.state.reserve.cols[outerIndex].pop()
                        break
                    case "work":
                        srcCol.pop()
                        if (srcCol.length > 0) {
                            // reveal any face down cards
                            srcCol[srcCol.length - 1] = srcCol[srcCol.length - 1] & ~cardutils.FLIP_FLAG
                        }
                        break
                }
                this.select("none")
                return {
                    raw: src,
                    prevOuterIndex: outerIndex,
                    prevInnerIndex: innerIndex,
                    foundationIndex: dstIndex,
                    prevFoundationRaw
                }
            }
        }
        return null
    }

    sendAny() {
        if (!this.canAutoWin()) {
            return null
        }
        for (let workIndex = 0; workIndex < this.state.work.length; workIndex++) {
            var work = this.state.work[workIndex]
            if (work.length > 0) {
                var sent = this.sendHome("work", workIndex, work.length - 1)
                if (sent != null) {
                    this.queueSave()
                    return sent
                }
            }
        }
        return null
    }

    // -----------------------------------------------------------------------------------------------
    // Selection

    select(type, outerIndex, innerIndex) {
        if (outerIndex == null) {
            outerIndex = 0
        }
        if (innerIndex == null) {
            innerIndex = 0
        }
        if (
            this.state.selection.type === type &&
            this.state.selection.outerIndex === outerIndex &&
            this.state.selection.innerIndex === innerIndex
        ) {
            // Toggle
            type = "none"
            outerIndex = 0
            innerIndex = 0
        }

        return (this.state.selection = {
            type,
            outerIndex,
            innerIndex
        })
    }

    eatSelection() {
        switch (this.state.selection.type) {
            case "pile":
                this.state.pile.cards.pop()
                break
            case "reserve":
                if (this.state.reserve != null) {
                    this.state.reserve.cols[this.state.selection.outerIndex].pop()
                }
                break
            case "work":
                var srcCol = this.state.work[this.state.selection.outerIndex]
                if (this.state.selection.innerIndex >= 0) {
                    while (this.state.selection.innerIndex < srcCol.length) {
                        srcCol.pop()
                    }
                }
                if (srcCol.length > 0) {
                    // reveal any face down cards
                    srcCol[srcCol.length - 1] = srcCol[srcCol.length - 1] & ~cardutils.FLIP_FLAG
                }
                break
        }
        return this.select("none")
    }

    getSelection() {
        const selectedCards = []
        switch (this.state.selection.type) {
            case "pile":
                if (this.state.pile.cards.length > 0) {
                    selectedCards.push(this.state.pile.cards[this.state.pile.cards.length - 1])
                }
                break
            case "reserve":
                if (this.state.reserve != null) {
                    const col = this.state.reserve.cols[this.state.selection.outerIndex]
                    if (col.length > 0) {
                        selectedCards.push(col[col.length - 1])
                    }
                }
                break
            case "work":
                if (this.state.selection.outerIndex >= 0 && this.state.selection.innerIndex >= 0) {
                    const srcCol = this.state.work[this.state.selection.outerIndex]
                    let index = this.state.selection.innerIndex
                    while (index < srcCol.length) {
                        selectedCards.push(srcCol[index])
                        index += 1
                    }
                }
                break
        }

        return selectedCards
    }

    // -----------------------------------------------------------------------------------------------
    // Tweens

    addTweens(tweens) {
        return this.app.addTweens(tweens)
    }
}

// -----------------------------------------------------------------------------------------------

export default SolitaireGame

File: tvos/Sources/Game/SolitaireGame.swift
import Foundation

// Direct port of src/SolitaireGame.js. Holds the current mode + state, the
// undo stack, persistence, and the generic helpers the modes call into.
final class SolitaireGame {
    var state: GameState!
    var modeId: String = "klondike"
    var hard: Bool = false // toggle for the *next* game; state.hard is this game's
    var undoStack: [GameState] = []

    let modeOrder = ["baker", "eagle", "emperor", "freecell", "golf", "klondike", "scorpion", "spider", "spiderette", "yukon"]
    let modes: [String: GameMode] = [
        "baker": BakerMode(),
        "eagle": EagleMode(),
        "emperor": EmperorMode(),
        "freecell": FreecellMode(),
        "golf": GolfMode(),
        "klondike": KlondikeMode(),
        "scorpion": ScorpionMode(),
        "spider": SpiderMode(),
        "spiderette": SpideretteMode(),
        "yukon": YukonMode(),
    ]

    var mode: GameMode { modes[modeId] ?? modes["klondike"]! }

    init() {
        if !load() {
            newGame()
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Persistence (UserDefaults stands in for the web build's localStorage)

    private struct SavePayload: Codable {
        var mode: String
        var hard: Bool
        var state: GameState
    }

    func load() -> Bool {
        guard let data = UserDefaults.standard.data(forKey: "save"),
              let payload = try? JSONDecoder().decode(SavePayload.self, from: data),
              modes[payload.mode] != nil
        else {
            return false
        }
        modeId = payload.mode
        hard = payload.hard
        state = payload.state
        undoStack = []
        return true
    }

    func save() {
        let payload = SavePayload(mode: modeId, hard: hard, state: state)
        if let data = try? JSONEncoder().encode(payload) {
            UserDefaults.standard.set(data, forKey: "save")
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Undo

    var canUndo: Bool { !undoStack.isEmpty }

    func pushUndo() {
        undoStack.append(state)
    }

    func undo() {
        if let prev = undoStack.popLast() {
            state = prev
            save()
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Generic input handlers

    func newGame(_ newMode: String? = nil) {
        if let newMode, modes[newMode] != nil {
            modeId = newMode
        }
        mode.newGame(self)
        undoStack = []
        save()
    }

    func click(_ type: SpotType, _ outerIndex: Int = 0, _ innerIndex: Int = 0, isRightClick: Bool = false) {
        let before = state!
        pushUndo()
        mode.click(self, type, outerIndex, innerIndex, isRightClick)
        // The web build snapshots every click, including pure selection
        // changes; on a remote that makes Undo feel broken (it "does
        // nothing" several times). Only keep snapshots of real board changes.
        if state.boardEquals(before) {
            _ = undoStack.popLast()
        }
        save()
    }

    func won() -> Bool {
        mode.won(self)
    }

    func lost() -> Bool {
        mode.lost(self)
    }

    // -----------------------------------------------------------------------------------------------
    // Generic helpers

    func findFoundationSuitIndex(_ raw: Int) -> Int {
        if state.foundations.isEmpty {
            return -1
        }

        // find a matching suit
        let srcInfo = CardUtils.info(raw)
        for (fIndex, f) in state.foundations.enumerated() where f >= 0 {
            let dstInfo = CardUtils.info(f)
            if srcInfo.suit == dstInfo.suit && (srcInfo.value == dstInfo.value + 1 || srcInfo.value == dstInfo.value - 12) {
                return fIndex
            }
        }

        // find a free slot
        for (fIndex, f) in state.foundations.enumerated() where f < 0 {
            return fIndex
        }

        return -1
    }

    func workPileEmpty() -> Bool {
        for work in state.work where !work.isEmpty {
            return false
        }
        return true
    }

    func reserveEmpty() -> Bool {
        guard let reserve = state.reserve else {
            return true
        }
        for col in reserve.cols where !col.isEmpty {
            return false
        }
        return true
    }

    func canAutoWin() -> Bool {
        if won() {
            return false
        }
        if state.foundations.isEmpty {
            return false
        }
        if !state.draw.cards.isEmpty {
            return false
        }
        if !state.pile.cards.isEmpty {
            return false
        }
        if let reserve = state.reserve {
            for col in reserve.cols where !col.isEmpty {
                return false
            }
        }
        for col in state.work {
            var last = 100
            for raw in col {
                if (raw & CardUtils.FLIP_FLAG) != 0 {
                    return false
                }
                let info = CardUtils.info(raw)
                if info.value > last {
                    return false
                }
                last = info.value
            }
        }
        return true
    }

    @discardableResult
    func sendHome(_ type: SpotType, _ outerIndex: Int, _ innerIndex: Int) -> Bool {
        var src: Int? = nil
        switch type {
        case .pile:
            if let last = state.pile.cards.last {
                src = last
            }
        case .reserve:
            if let reserve = state.reserve, reserve.cols.count > outerIndex, let last = reserve.cols[outerIndex].last {
                src = last
            }
        case .work:
            let srcCol = state.work[outerIndex]
            if innerIndex != srcCol.count - 1 {
                return false
            }
            if innerIndex >= 0 && innerIndex < srcCol.count {
                src = srcCol[innerIndex]
            }
        default:
            break
        }

        guard let src else {
            return false
        }

        let srcInfo = CardUtils.info(src)
        let dstIndex = findFoundationSuitIndex(src)
        if dstIndex >= 0 {
            var sendHome = false
            if state.foundations[dstIndex] >= 0 {
                let dstInfo = CardUtils.info(state.foundations[dstIndex])
                if srcInfo.value == dstInfo.value + 1 || srcInfo.value == dstInfo.value - 12 {
                    sendHome = true
                }
            } else {
                let foundationBase = state.foundationBase ?? 0 // Ace
                if srcInfo.value == foundationBase {
                    sendHome = true
                }
            }

            if sendHome {
                state.foundations[dstIndex] = src
                switch type {
                case .pile:
                    state.pile.cards.removeLast()
                case .reserve:
                    state.reserve?.cols[outerIndex].removeLast()
                case .work:
                    state.work[outerIndex].removeLast()
                    revealTopOfWork(outerIndex)
                default:
                    break
                }
                select(.none)
                return true
            }
        }
        return false
    }

    @discardableResult
    func sendAny() -> Bool {
        if !canAutoWin() {
            return false
        }
        for workIndex in 0..<state.work.count {
            let work = state.work[workIndex]
            if !work.isEmpty {
                if sendHome(.work, workIndex, work.count - 1) {
                    save()
                    return true
                }
            }
        }
        return false
    }

    // Reveal any face down card left on top of a work column
    func revealTopOfWork(_ outerIndex: Int) {
        let count = state.work[outerIndex].count
        if count > 0 {
            state.work[outerIndex][count - 1] = state.work[outerIndex][count - 1] & ~CardUtils.FLIP_FLAG
        }
    }

    // -----------------------------------------------------------------------------------------------
    // Selection

    func select(_ type: SpotType, _ outerIndex: Int = 0, _ innerIndex: Int = 0) {
        var type = type
        var outerIndex = outerIndex
        var innerIndex = innerIndex
        if state.selection.type == type && state.selection.outerIndex == outerIndex && state.selection.innerIndex == innerIndex {
            // Toggle
            type = .none
            outerIndex = 0
            innerIndex = 0
        }

        state.selection = Selection(type: type, outerIndex: outerIndex, innerIndex: innerIndex)
    }

    func eatSelection() {
        switch state.selection.type {
        case .pile:
            state.pile.cards.removeLast()
        case .reserve:
            let outer = state.selection.outerIndex
            state.reserve?.cols[outer].removeLast()
        case .work:
            let outer = state.selection.outerIndex
            if state.selection.innerIndex >= 0 {
                while state.selection.innerIndex < state.work[outer].count {
                    state.work[outer].removeLast()
                }
            }
            revealTopOfWork(outer)
        default:
            break
        }
        select(.none)
    }

    func getSelection() -> [Int] {
        var selectedCards: [Int] = []
        switch state.selection.type {
        case .pile:
            if let last = state.pile.cards.last {
                selectedCards.append(last)
            }
        case .reserve:
            if let reserve = state.reserve {
                let col = reserve.cols[state.selection.outerIndex]
                if let last = col.last {
                    selectedCards.append(last)
                }
            }
        case .work:
            if state.selection.outerIndex >= 0 && state.selection.innerIndex >= 0 {
                let srcCol = state.work[state.selection.outerIndex]
                var index = state.selection.innerIndex
                while index < srcCol.count {
                    selectedCards.append(srcCol[index])
                    index += 1
                }
            }
        default:
            break
        }

        return selectedCards
    }

    // -----------------------------------------------------------------------------------------------
    // Shared mode helpers (the common draw/foundation click logic repeated in
    // most of the JS modes)

    // klondike/baker/freecell-style draw click: recycle the pile back into the
    // draw when allowed, otherwise flip `pile.show` cards onto the pile.
    // `useRedeals` is the eagle variant where recycles are counted.
    func standardDrawClick(allowRecycle: Bool, useRedeals: Bool = false) {
        if allowRecycle && state.draw.cards.isEmpty {
            if useRedeals {
                if (state.draw.redeals ?? 0) > 0 {
                    state.draw.redeals = (state.draw.redeals ?? 0) - 1
                    state.draw.cards = state.pile.cards
                    state.pile.cards = []
                }
            } else {
                state.draw.cards = state.pile.cards
                state.pile.cards = []
            }
        } else {
            var cardsToDraw = state.pile.show
            if cardsToDraw > state.draw.cards.count {
                cardsToDraw = state.draw.cards.count
            }
            for _ in 0..<cardsToDraw {
                state.pile.cards.append(state.draw.cards.removeFirst())
            }
        }
        select(.none)
    }

    // The "gauntlet of breaks" foundation click shared by klondike, baker,
    // emperor, freecell, yukon (base 0, no wrap) and eagle (foundationBase,
    // wrapping king to ace).
    func standardFoundationClick(_ outerIndex: Int, base: Int = 0, wrap: Bool = false) {
        let src = getSelection()
        gauntlet: while true {
            if src.count != 1 {
                break gauntlet
            }
            let srcInfo = CardUtils.info(src[0])
            if state.foundations[outerIndex] < 0 {
                // empty
                if srcInfo.value != base {
                    break gauntlet
                }
            } else {
                let dstInfo = CardUtils.info(state.foundations[outerIndex])
                if srcInfo.suit != dstInfo.suit {
                    break gauntlet
                }
                if srcInfo.value != dstInfo.value + 1 && !(wrap && srcInfo.value == dstInfo.value - 12) {
                    break gauntlet
                }
            }

            state.foundations[outerIndex] = src[0]
            eatSelection()
            break gauntlet
        }
        select(.none)
    }
}

// Concatenates `copies` copies of cards, tagging each repeat with COPY bits so
// duplicate cards stay distinguishable (see CardUtils.COPY_MASK).
func deckCopies(_ cards: [Int], _ copies: Int) -> [Int] {
    var deck: [Int] = []
    for copy in 0..<copies {
        deck.append(contentsOf: cards.map { $0 | (copy << CardUtils.COPY_SHIFT) })
    }
    return deck
}

func shuffled(_ array: [Int]) -> [Int] {
    var array = array
    for i in stride(from: array.count - 1, to: 0, by: -1) {
        let j = Int.random(in: 0...i)
        array.swapAt(i, j)
    }
    return array
}
